woodchipper/http/fastapi.py

```diff
@@ -0,0 +1,73 @@
+import uuid
+from typing import Any, Callable, Coroutine
+
+from fastapi import FastAPI, Request, Response
+from fastapi.routing import APIRoute
+
+from woodchipper.context import LoggingContext, logging_ctx
+
+BLACKLISTED_HEADERS = ["authorization", "cookie"]
+
+
+class WoodchipperFastAPI:
+    def __init__(self, app: FastAPI, blacklisted_headers=BLACKLISTED_HEADERS, request_id_factory=None):
+        self._app = app
+        self._blacklisted_headers = blacklisted_headers
+        self._request_id_factory = request_id_factory or (lambda: str(uuid.uuid4()))
+
+    def get_custom_route_handler(self, gen_id, blacklisted_headers):
+        def get_route_handler(self) -> Callable[[Request], Coroutine[Any, Any, Response]]:
+
+            original_route_handler = super(type(self), self).get_route_handler()
+
+            async def woodchipper_route_handler(request: Request) -> Response:
+
+                # FastAPI doesn't handle this conversion until after the routehandler is called so we
+                # have to do it here.
+                queries = {}
+                for k, v in request.query_params.multi_items():
+                    if k in queries and not isinstance(queries[k], list):
+                        queries[k] = [queries[k], v]
+                    elif k in queries:
+                        queries[k].append(v)
+                    else:
+                        queries[k] = v
```

Comment:
Why the checks for entries already in `queries`?

Reply:
The request object doesn't combine repeat query params into a list. They live as separate entries in the multi_items attribute. This creates a queries representation that turns them into a list for the logging context.

You can explore it more if you dig into here:
https://github.com/encode/starlette/blob/fa737beb2838f568fe2477e2316a7fdf3d3eb257/starlette/datastructures.py#L386